File: decision_tool_tester/src/GlobalProfileTester.cpp
#include "ros/ros.h"
#include "ros/package.h"
#include "DecisionTool.h"
#include <string>

using namespace dt;
using namespace std;

void print_vector(vector<string> &v, bool type);

int main(int argc, char **argv)
{
	/* Initialize ROS */
	ros::init(argc, argv, "DecisionToolTester");

	/* Initialize ROS node - useful for ROS_INFO etc. */
	ros::NodeHandle n;

	/* Get path to testfiles (path is for DecisionService node) */
	std::string path = ros::package::getPath("decision_service").append("/testfiles/");

	DecisionTool *dta = new DecisionTool;

	/* Create a new engine for a particular component and load the respective ruleset */
	if(dta->EngineExists("firstid")) {
		dta->DisposeEngine("firstid");
	}
	if(!dta->CreateEngine("firstid", path +"example3.txt")) {
		delete dta;
		return 1;
	}

	/* Create a new engine for another component and load the SAME ruleset */
	if(dta->EngineExists("secondid")) {
		dta->DisposeEngine("secondid");
	}
	if(!dta->CreateEngine("secondid", path+"example3.txt")) {
		delete dta;
		return 1;
	}

	/* Set global parameter (male/female) */
	dta->SetGlobalConditional("user.male", true);

	/* The first engine will use capital letters, while the second will not */
	dta->SetConditional("firstid", "response.capslock", true);
	dta->SetConditional("secondid", "response.capslock", false);

	/* Evaluate WhatAmI for each engine */
	printf("What am I, first engine?  : %s\n", dta->Evaluate("firstid", "WhatAmI").c_str());
	printf("What am I, second engine? : %s\n", dta->Evaluate("secondid", "WhatAmI").c_str());

	/* Dispose the first engine */
	dta->DisposeEngine("firstid");

	/* Dispose the second engine */
	dta->DisposeEngine("secondid");
	
	/* Playing with global attributes (names do not follow conventions for demo purposes) */
	dta->SetGlobalConditional("hobbit.voice.volume", 7);
	dta->SetGlobalConditional("hobbit.voice.male", false);
	dta->SetGlobalConditional("hobbit.voice.pitch", "high");

	/* Just test attributes to check if the prefix services work */
	dta->SetGlobalConditional("hobbit.name", "George");
	dta->SetGlobalConditional("house.temperature", 27.2);

	/* Test engine */
	if(dta->EngineExists("test")) {
		dta->DisposeEngine("test");
	}
	if(!dta->CreateEngine("test", path +"example1.txt")) {
		delete dta;
		return 1;
	}
	dta->SetConditional("test", "hobbit.nickname", "Bob");
	dta->SetConditional("test", "house.size", "Big");

	/* Print what each profile contains at this point */
	printf("\nGlobal profile:\nhobbit.voice.volume = 7\nhobbit.voice.male = false\nhobbit.voice.pitch = high\n");
	printf("hobbit.name = George\nhouse.temperature = 27.2\n\n");
	printf("Local parameters for engine \"test\":\nhobbit.nickname = Bob\nhouse.size = Big\n\n");

	/* Get attributes based on prefix */
	printf("Getting local attributes for \"test\" and global attributes beginning with \"house.\" (attribute type included):\n");
	vector<string> v = dta->GetProfileAttributesWithPrefix("test", "house.*", true);
	print_vector(v, true);

	printf("Getting local attributes for \"test\" and global attributes beginning with \"ho\" (attribute type included):\n");
	v = dta->GetProfileAttributesWithPrefix("test", "ho*", true);
	print_vector(v, true);

	printf("Getting global attributes beginning with \"hobbit.\"(attribute type NOT included):\n");
	v = dta->GetProfileAttributesWithPrefix("", "hobbit.*", false);
	print_vector(v, false);

	printf("Getting global attributes beginning with \"hobbit.\" (attribute type NOT included):\n");
	//alternative way to achieve the same result as before
	v = dta->GetGlobalProfileAttributesWithPrefix("hobbit.*");
	print_vector(v, false);

	printf("Getting global attributes beginning with \"ho\" (attribute type NOT included):\n");
	v = dta->GetGlobalProfileAttributesWithPrefix("ho*");
	print_vector(v, false);

	dta->DisposeEngine("test");

	delete dta;

	/* Shut down ROS */
	ros::shutdown();

	return 0;
}

void print_vector(vector<string> &v, bool type) {
	vector<string>::iterator it;
	for(it = v.begin(); it != v.end(); ++it) {
		printf("%s\n", (*it).c_str());
		++it;
		if(type) {
			printf("\tType:\t%s\n", (*it).c_str());
			++it;
		}
		printf("\tValue:\t%s\n", (*it).c_str());
	}
	printf("\n==========\n");
	v.clear();
}

File: decision_tool_tester/src/decision_tool_tester/GlobalProfileTester.py
#!/usr/bin/env python

import roslib
import rospy
from decision_service.decision_tool import DecisionTool

def main():
    # id for first engine
    firstid = 'id1'
    
    # id for second engine
    secondid = 'id2'
    
    # Get path to testfiles (path is for DecisionService node)
    path = roslib.packages.get_pkg_dir('decision_service') + '/testfiles/'
    
    dt = DecisionTool()
    # Create a new engine for a particular component and load the respective ruleset
    if dt.EngineExists(firstid):
        dt.DisposeEngine(firstid)
    dt.CreateEngine(firstid, path + 'example3.txt')
    
    # Create a new engine for another component and load the SAME ruleset
    if dt.EngineExists(secondid):
        dt.DisposeEngine(secondid)
    dt.CreateEngine(secondid, path + 'example3.txt')
    
    # Set global parameter (male/female)
    dt.SetGlobalConditional('user.male', True)
    
    # The first engine will use capital letters, while the second will not
    dt.SetConditional(firstid, 'response.capslock', True)
    dt.SetConditional(secondid, 'response.capslock', False)
    
    # Evaluate WhatAmI for each engine
    print 'What am I, first engine?  : ' + dt.Evaluate(firstid, 'WhatAmI')
    print 'What am I, second engine?  : ' + dt.Evaluate(secondid, 'WhatAmI')
    
    # Dispose the first engine
    dt.DisposeEngine(firstid)
    
    # Dispose the second engine
    dt.DisposeEngine(secondid)
    
    # Playing with global attributes (names do not follow conventions for demo purposes)
    dt.SetGlobalConditional('hobbit.voice.volume', 7)
    dt.SetGlobalConditional('hobbit.voice.male', False)
    dt.SetGlobalConditional('hobbit.voice.pitch', 'high')
    
    # Just test attributes to check if the prefix services work
    dt.SetGlobalConditional('hobbit.name', 'George')
    dt.SetGlobalConditional('house.temperature', 27.2)
    
    # Test engine
    if dt.EngineExists('test'):
        dt.DisposeEngine('test')
    dt.CreateEngine('test', path + 'example1.txt')
    dt.SetConditional('test', 'hobbit.nickname', 'Bob')
    dt.SetConditional('test', 'house.size', 'Big')
    
    # Print what each profile contains at this point
    print '\nGlobal profile:\nhobbit.voice.volume = 7\nhobbit.voice.male = false\nhobbit.voice.pitch = high'
    print 'hobbit.name = George\nhouse.temperature = 27.2\n'
    print 'Local parameters for engine \"test\":\nhobbit.nickname = Bob\nhouse.size = Big\n'
    
    # Get attributes based on prefix
    print 'Getting local attributes for \"test\" and global attributes beginning with \"house.\":'
    print dt.GetProfileAttributesWithPrefix('test', 'house.*')
    
    print 'Getting local attributes for \"test\" and global attributes beginning with \"ho\":'
    print dt.GetProfileAttributesWithPrefix('test', 'ho*')
    
    print 'Getting global attributes beginning with \"hobbit.\":'
    # alternative way to achieve the same result as before
    print dt.GetGlobalProfileAttributesWithPrefix('hobbit.*')
    
    dt.DisposeEngine('test')
    
if __name__ == "__main__":
    main()
